util/grpc/interceptor.go

```diff
@@ -50,6 +58,38 @@ var (
 	}
 )
 
+// SetVersionHeaderUnaryServerInterceptor returns a new unary server interceptor that sets the argo-version header
+func SetVersionHeaderUnaryServerInterceptor(version wfv1.Version) grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		err := grpc.SetHeader(ctx, metadata.Pairs(ArgoVersionHeader, version.Version))
```

Comment:
another major difference with a header is its security properties; the version from `/info` is only exposed to authenticated clients. we'd have to set this only if the request is authenticated

Reply:
I mentioned this in the PR description, but the header won't be added for 401 errors, because the `as.gatekeeper.UnaryServerInterceptor()` is added before this one:
https://github.com/MasonM/argo-workflows/blob/5cff94056bd25bb9d961342de614ac8abaccd0aa/server/apiserver/argoserver.go#L306

Since `as.gatekeeper.UnaryServerInterceptor()` returns an error on an auth failure, that should prevent this interceptor from running, unless I'm misunderstanding how this works. I tested this locally:
```
$ curl -v http://localhost:2746/api/v1/info
* Host localhost:2746 was resolved.
* IPv6: ::1
* IPv4: 127.0.0.1
*   Trying [::1]:2746...
* connect to ::1 port 2746 from ::1 port 60801 failed: Connection refused
*   Trying 127.0.0.1:2746...
* Connected to localhost (127.0.0.1) port 2746
> GET /api/v1/info HTTP/1.1
> Host: localhost:2746
> User-Agent: curl/8.8.0
> Accept: */*
>
* Request completely sent off
< HTTP/1.1 401 Unauthorized
< Content-Type: application/json
< Trailer: Grpc-Trailer-Content-Type
< X-Ratelimit-Limit: 1000
< X-Ratelimit-Remaining: 999
< X-Ratelimit-Reset: Sat, 21 Sep 2024 23:29:45 UTC
< Date: Sat, 21 Sep 2024 23:29:44 GMT
< Transfer-Encoding: chunked
<
* Connection #0 to host localhost left intact
{"code":16,"message":"token not valid. see https://argo-workflows.readthedocs.io/en/latest/faq/"}
```

Reply:
Oh nice, thanks for confirming and testing that!

And sorry you are catching all my mistakes in review on this PR 😅 I was having a smoldering migraine all last week, so my brain is probably still adjusting (that's also why I didn't approve much last week)

Reply:
This is actually an excellently detailed PR description that answers all my questions 👏  that's also coming from the person who tends to write the most detailed descriptions in this repo 🙂

Reply:
I'm wondering if might be worth adding an E2E test to ensure there's no security regressions with this in the future -- what do you think?

Reply:
It's good you asked for that, because in the process of writing that test I realized that the header actually is being returned on 401s when `--auth-mode=client` is used. Earlier, I had only tested with `--auth-mode=sso`, which causes `as.gatekeeper.UnaryServerInterceptor()` to immediately error out on auth failures: https://github.com/argoproj/argo-workflows/blob/6cac182ff247ff37a13b8efc6cb4ff11fcdd20c3/server/auth/gatekeeper.go#L207-L210

But with `--auth-mode=client`. it doesn't actually verify the token is valid, just that it's well-formed. It's only later in the request lifecycle (when the token is used) that you can get the 401.

I pushed f590dcb0b754e95f57c698f3c3d7dcf7d43e68ab to change the interceptors to be post-processors, and only set the header if there isn't an error. Of course, I also updated the E2E tests.

Reply:
Yes that's correct regarding `--auth-mode=client`. And yes I was trying to make sure it was a post-processor with my limited knowledge of `grpc-go`. I didn't realize there was a timing difference to the code though (just commented on the general area); fortunately you figured it out eventually so at least now that's properly covered 😅